Order Details > Shipping Label: reprint shipping label main screen

jaclync opened the pull request:
Part of #2169 

For code reviewers: please review https://github.com/woocommerce/woocommerce-ios/pull/3341 first. Feel free to hold off reviewing this PR if the base PR has changes requested.

## Why

This PR implements the main screen for reprinting a shipping label. The view model is implemented in the base PR https://github.com/woocommerce/woocommerce-ios/pull/3341 since the UI changes are large enough.

## Changes

- Added a new icon `pagesFootnoteImage` for the footnote size for `pages` Gridicon
- Added an extension `UIButton.applyFootnoteLinkButtonStyle` tot apply footnote link button styles (footnote has smaller font size)
- Added an extension `UITableViewCell.hideSeparator` to hide the separator for a cell - feel free to suggest other solutions!
- Created `ReprintShippingLabelViewController` with a table view and a CTA to reprint a shipping label. I decided to implement the UI above the CTA with table view cells, after struggling with dynamic height using `tableView.tableHeaderView`.)
- Created `SpacerTableViewCell` for a cell just for vertical padding
- Integration with order details in `OrderDetailsViewController` and `OrderDetailsDataSource`

## Testing

Prerequisites: the site has [WooCommerce Shipping & Tax plugin](https://docs.woocommerce.com/document/woocommerce-shipping-and-tax/) and there is an order with at least one non-refunded shipping label.

You can set up your store to use a test credit card with instructions in p91TBi-3xD-p2.

- Go to the orders tab
- Tap on an order with at least one non-refunded shipping label
- After syncing, tap on "Reprint Shipping Label" CTA --> it should navigate to the reprint UI as in screenshots

## Example screenshots

Notes for reviewers: due to the current implementation of `TopLeftImageTableViewCell` still using frames and bounds, the images in "See layout and paper sizes options" and "Don't know how to print from your device?" rows are not the footnote size 20x20 as in design. I added this to the design enhancements issue https://github.com/woocommerce/woocommerce-ios/issues/3334 to work on them separately.

light | dark
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-12-14 at 13 58 18](https://user-images.githubusercontent.com/1945542/102047199-cf6b0780-3e17-11eb-869c-8a0f1cecfb1c.png) | ![Simulator Screen Shot - iPhone 11 - 2020-12-14 at 14 24 57](https://user-images.githubusercontent.com/1945542/102047401-2cff5400-3e18-11eb-8dfe-46db86007fc9.png)



Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jaclync committed:
Create `ReprintShippingLabelViewController` that has a top `UIScrollView` but it turns out to be hard for the paper size row.

jaclync committed:
Add `ShippingLabelAction.loadShippingLabelSettings` to load settings for a shipping label.

jaclync committed:
Add an extension helper to `ShippingLabelPaperSize` for UI display description text.

jaclync committed:
Create `ReprintShippingLabelHeaderView` for static header view in reprint shipping label screen.

jaclync committed:
Create `SpacerTableViewCell` for vertical spacing between cells.

jaclync committed:
Implement reprint UI with table view cells.

jaclync committed:
Remove unused param for `ReprintShippingLabelViewController` and navigate to `ReprintShippingLabelViewController` from order details.

jaclync committed:
Remove unused files `ReprintShippingLabelHeaderView` in favor of reusable cells (many resizing issues from tableHeaderView).

jaclync committed:
Move data-related logic from `ReprintShippingLabelViewController` to `ReprintShippingLabelViewModel` with unit tests.

jaclync committed:
Add test cases for `ShippingLabelAction.loadShippingLabelSettings`.

jaclync committed:
Delete classes that are coming from a different branch.

jaclync committed:
Merge branch 'issue/2169-load-ShippingLabelSettings-for-paper-size' into issue/2169-reprint-shipping-label-screen

jaclync committed:
Create and use footnote icon for  `pages` Gridicon.

jaclync committed:
Add more comments before PR.

jaclync committed:
Merge branch 'issue/2169-load-ShippingLabelSettings-for-paper-size' into issue/2169-reprint-shipping-label-screen

jaclync committed:
Update to `@Published` for view model's `selectedPaperSize`.

Garance91540 reviewed (approved):
Hi @jaclync it looks good but slightly too dense, could we update the padding under the info text "if you already used..." from 16px to 24px? Thank you!

shiki commented on WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    deinit {
+        cancellables.forEach {
+            $0.cancel()
+        }
```
I still have to validate this myself but we probably won't need to manually call `cancel()` on `deinit`. From [the docs](https://developer.apple.com/documentation/combine/anycancellable):

> An AnyCancellable instance automatically calls cancel() when deinitialized.

This is as long as we're not capturing `self` in the `sink()` calls. 

---

Nothing to change here though. I just thought of mentioning it for the sake of sharing ideas. 🙂

jaclync replied:
Interesting! I checked the memory debugger before & after removing these lines, and saw 2 objects with `reprint` keyword in the memory after leaving the reprint screen in both cases:

<img width="807" alt="Screen Shot 2020-12-15 at 11 55 01 AM" src="https://user-images.githubusercontent.com/1945542/102169274-cee07880-3ecc-11eb-8977-81beb9681ef6.png">

These two objects weren't marked with memory warning, and there were other `Swift.ReferenceWritableKeyPath` as well. I didn't find much information online other than the [use of `assign(to:)` in the Swift forum](https://forums.swift.org/t/does-assign-to-produce-memory-leaks/29546). We also weak self in the `sink` closure, not sure if there's any way to get rid of the key paths.

I see that we still cancel `AnyCancellable` manually in other places in the codebase, I'll leave them for now and we can remove them if we feel like in the future.

shiki replied:
OMG the documentation lies. Lies!!! 

Thanks for looking into it!

jaclync replied:
These `Swift.ReferenceWritableKeyPath`s were also alive after manually cancelling the `AnyCancellable` so I'm not sure what to do there 😓 at least there are no memory warnings on these 🙄

shiki commented on WooCommerce/Classes/Extensions/UIButton+Helpers.swift:
```diff
@@ -80,6 +80,17 @@ extension UIButton {
+        setTitleColor(.systemColor(.secondaryLabel), for: .normal)
+        setTitleColor(.systemColor(.tertiaryLabel), for: .highlighted)
+    }
+
```
⚠️  This does not seem to be used anywhere. 🙂 

```suggestion

```

jaclync replied:
Wow nice catch 🦅 👁️ ! (This is from my iteration of using a custom view for `table.tableFooterView` but the dynamic height is tricky after rotation.) Removed in https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/341def7b4e3b1c75abc921c16ac423e0dd4945a6

shiki replied:
🤸

shiki commented on WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift:
```diff
@@ -0,0 +1,257 @@
+extension ReprintShippingLabelViewController: UITableViewDataSource {
+    func numberOfSections(in tableView: UITableView) -> Int {
+        1
+    }
```
⚠️  I believe the default is 1. 

```suggestion

```

jaclync replied:
Ah right! Removed in https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/341def7b4e3b1c75abc921c16ac423e0dd4945a6

shiki commented on WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift:
```diff
@@ -0,0 +1,257 @@
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let row = rows[indexPath.row]
```
If there's one thing I hate, it's “invalid number of rows” exceptions. 😄 I think we should still consider using `safe` whenever possible. 

```suggestion
        guard let row = rows[safe: indexPath.row] else {
            return UITableViewCell()
        }
```

But up to you. It's unlikely. We can always fix it if we ever get a crash later on the road.

jaclync replied:
Hah right, it doesn't hurt to be safer. Updated in: https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/79499744b164335f3b80cad620152502c4d1a96b

shiki commented on WooCommerce/Classes/ViewRelated/ReusableViews/SpacerTableViewCell.swift:
```diff
@@ -0,0 +1,24 @@
+/// The cell currently has default cell background color (`.listForeground`).
+final class SpacerTableViewCell: UITableViewCell {
+    /// The height of the spacer cell.
+    var height: CGFloat = 0 {
+        didSet {
+            heightConstraint.constant = height
+        }
+    }
```
Do you think we should consider `configure(height:)` instead so we're consistent (kind of) with the other cells?

jaclync replied:
yup makes sense! updated in: https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/79499744b164335f3b80cad620152502c4d1a96b

shiki commented on WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift:
```diff
@@ -0,0 +1,257 @@
+        configureCommonStylesForInfoLinkCell(cell)
+    }
+
+    func configureCommonStylesForInfoLinkCell(_ cell: TopLeftImageTableViewCell) {
+        cell.imageView?.tintColor = .systemColor(.secondaryLabel)
+        cell.textLabel?.applyFootnoteStyle()
+        cell.textLabel?.textColor = .systemColor(.secondaryLabel)
+        cell.hideSeparator()
+        cell.selectionStyle = .default
+    }
```
ℹ️  I think we should consider this as a legitimate configuration style for `TopLeftImageTableViewCell`. Something like:

```swift
cell.configure(style: .secondaryOrWhatever)
```

This is similar to what we did with [`HeadlineLabelTableViewCell`](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/ViewRelated/ReusableViews/HeadlineLabelTableViewCell.swift):

https://github.com/woocommerce/woocommerce-ios/blob/b8e1d3b3b4b06b8a13b988ef98c64c0a37481f65/WooCommerce/Classes/ViewRelated/ReusableViews/HeadlineLabelTableViewCell.swift#L53-L66

The theory is that:

- We have standard “styles” to use all throughout the app.
- ViewControllers would only be using styles and not defining their own standards.
- Probably makes migration to SwiftUI easier. Probably. 

Something to think about in the future. 🙂

jaclync replied:
Great suggestion! It is nice to have a set of predefined styles. So far, I just see the consistent usage of font size for `TopLeftImageTableViewCell ` while the image/text colors differ in use cases. Updated in: https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/830f6aede4fbbd41467f01c0b8bbe579831d0871

shiki commented on WooCommerce/Classes/ViewModels/Order Details/Shipping Labels/ReprintShippingLabelViewController.swift:
```diff
@@ -0,0 +1,257 @@
+
+private extension ReprintShippingLabelViewController {
+    enum Constants {
+        static let cellValueTextColor = UIColor.systemColor(.secondaryLabel)
```
⚠️ This seems to be unused.

```suggestion

```

jaclync replied:
thanks for catching this! I wish Xcode has an unused code highlighter. Removed in https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/341def7b4e3b1c75abc921c16ac423e0dd4945a6

shiki commented on WooCommerce/Classes/ViewRelated/ReusableViews/SpacerTableViewCell.xib:
```diff
@@ -0,0 +1,49 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I'm curious if you've tried not having a XIB. Will it work? It seems unnecessary. 😄 This is nitpicky so feel free to ignore.

jaclync replied:
I thought about not using a xib for 1 subview, though I generally still create a xib whenever possible in Xcode (`UITableViewHeaderFooterView` does not have a xib option in Xcode new file window) following our practices in the codebase. Also, our table view cell registration usually expects the use of a xib (`tableView.registerNib(for:)`) so I thought it's easier not to create exception for table view configuration 😅 .

Xib does create tens of XML diffs even it just has one subview 🙄 looking forward to gradually moving to SwiftUI!

shiki reviewed (approved):
Looking good! I just have a few suggestions for consideration (⚠️ ). 


---

> Added a new icon pagesFootnoteImage for the footnote size for pages Gridicon

GridIcon please don't crash. 🙈 

---

### Accessibility

It looks like there some issues with larger font sizes. 

The icon is misaligned with the text:

<img src="https://user-images.githubusercontent.com/198826/102104000-e9422580-3dea-11eb-9563-a0f89abbf596.png" width="300">

And the Paper Size does not scale very well. 

<img src="https://user-images.githubusercontent.com/198826/102104138-0f67c580-3deb-11eb-9255-ffafae80d9e3.png" width="300">

We can probably tackle these as separate issues. 🙂

jaclync committed:
Merge branch 'issue/2169-load-ShippingLabelSettings-for-paper-size' into issue/2169-reprint-shipping-label-screen

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

jaclync committed:
Remove unused code.

jaclync committed:
Improvements based on CR feedback.

jaclync committed:
Add `Style` enum to `TopLeftImageTableViewCell` for different font size configurations.

jaclync committed:
Add 8px spacer below info text row.

jaclync commented:
Thank you for the review & feedback @shiki and @Garance91540! I responded to the comments and updated the PR for another look 😃 

> Hi @jaclync it looks good but slightly too dense, could we update the padding under the info text "if you already used..." from 16px to 24px? Thank you!

Hi @Garance91540 yep! I added another 8px-tall spacer row under the info text in https://github.com/woocommerce/woocommerce-ios/pull/3342/commits/f696b7f6b743a85c4cf35178f867fd45a2c08f11 (cc @shiki) and here's the comparison:

before | after
-- | --
![Simulator Screen Shot - iPhone 11 - 2020-12-15 at 11 26 08](https://user-images.githubusercontent.com/1945542/102169745-da806f00-3ecd-11eb-836a-550cb8920db5.png) | ![Simulator Screen Shot - iPhone 11 - 2020-12-15 at 11 27 23](https://user-images.githubusercontent.com/1945542/102169750-de13f600-3ecd-11eb-8bef-cf600cd0a0a5.png)

jaclync committed:
Create `ShippingLabelPaperSizeListSelectorCommand` for paper size list selector data provider.

jaclync committed:
ReprintShippingLabelViewModel: handle selected action from paper size list selector.

jaclync committed:
Add unit tests for `ShippingLabelPaperSizeListSelectorCommand`.

shiki reviewed (approved):
Still LGTM! I reviewed the added `ShippingLabelPaperSizeListSelectorCommand` too. 

I forgot to mention yesterday, thank you for adding lots of comments! 🙏

jaclync committed:
Revert "Add unit tests for `ShippingLabelPaperSizeListSelectorCommand`."

This reverts commit 72ebf7931f2ba315a6b900922c0bb298fa5cf606.

jaclync committed:
Revert "ReprintShippingLabelViewModel: handle selected action from paper size list selector."

This reverts commit adb2b0a782ead972eaacf8bb9b534ca7487aa6af.

jaclync committed:
Revert "Create `ShippingLabelPaperSizeListSelectorCommand` for paper size list selector data provider."

This reverts commit e28ef2469cc7cf19a77edfedd25b01de8c67385b.

jaclync commented:
Sorry @shiki about the 3 commits on `ShippingLabelPaperSizeListSelectorCommand` by mistake! I decided to revert them instead of force pushing the branch because I remember from Platform9 that it is best not to alter the git history. Lemme know if you think force pushing would be cleaner 🙂

shiki reviewed (approved):
I have no opinion on force push or revert. 🤷 

:shipit:

jaclync commented:
Thank you for the review Jayson 💟 Merging this now for other dependent subtasks!
